File: boot/kernel_panic.c
/*
 * SlopOS Kernel Panic Handler
 * Emergency error handling for critical kernel failures
 * Uses serial driver for reliable output during panic situations
 */

#include <stdint.h>
#include "shutdown.h"
#include "../drivers/serial.h"
#include "../drivers/fate.h"
#include "../drivers/wl_currency.h"
#include "../video/roulette.h"
#include "../lib/numfmt.h"
#include "../lib/cpu.h"

static void panic_output_string(const char *str) {
    serial_emergency_puts(str);
}

/*
 * Get current instruction pointer for debugging
 */
static uint64_t get_current_rip(void) {
    uint64_t rip;
    __asm__ volatile ("leaq (%%rip), %0" : "=r" (rip));
    return rip;
}

/*
 * Get current stack pointer for debugging
 */
static uint64_t get_current_rsp(void) {
    uint64_t rsp;
    __asm__ volatile ("movq %%rsp, %0" : "=r" (rsp));
    return rsp;
}

/*
 * Main kernel panic routine
 * Displays error information and halts the system
 */
void kernel_panic(const char *message) {
    // Disable interrupts immediately
    cpu_cli();

    // Output panic header
    panic_output_string("\n\n");
    panic_output_string("=== KERNEL PANIC ===\n");

    // Output panic message
    if (message) {
        panic_output_string("PANIC: ");
        panic_output_string(message);
        panic_output_string("\n");
    } else {
        panic_output_string("PANIC: No message provided\n");
    }

    // Output debugging information
    panic_output_string("RIP: ");
    serial_emergency_put_hex(get_current_rip());
    panic_output_string("\n");

    panic_output_string("RSP: ");
    serial_emergency_put_hex(get_current_rsp());
    panic_output_string("\n");

    // Output CPU state information
    uint64_t cr0, cr3, cr4;
    __asm__ volatile ("movq %%cr0, %0" : "=r" (cr0));
    __asm__ volatile ("movq %%cr3, %0" : "=r" (cr3));
    __asm__ volatile ("movq %%cr4, %0" : "=r" (cr4));

    panic_output_string("CR0: ");
    serial_emergency_put_hex(cr0);
    panic_output_string("\n");

    panic_output_string("CR3: ");
    serial_emergency_put_hex(cr3);
    panic_output_string("\n");

    panic_output_string("CR4: ");
    serial_emergency_put_hex(cr4);
    panic_output_string("\n");

    panic_output_string("===================\n");
    panic_output_string("Skill issue lol\n");
    panic_output_string("System halted.\n");

    /*
     * Invoke the final purification ritual before shutdown
     * Paint all memory with 0x69—the essence of slop itself
     */
    execute_kernel();

    kernel_shutdown(message ? message : "panic");
}

/*
 * Kernel panic with additional context information
 */
void kernel_panic_with_context(const char *message, const char *function,
                              const char *file, int line) {
    // Disable interrupts immediately
    cpu_cli();

    panic_output_string("\n\n");
    panic_output_string("=== KERNEL PANIC ===\n");

    if (message) {
        panic_output_string("PANIC: ");
        panic_output_string(message);
        panic_output_string("\n");
    }

    if (function) {
        panic_output_string("Function: ");
        panic_output_string(function);
        panic_output_string("\n");
    }

    if (file) {
        panic_output_string("File: ");
        panic_output_string(file);
        if (line > 0) {
            panic_output_string(":");
            char line_buf[32];
            if (numfmt_u64_to_decimal((uint64_t)line, line_buf, sizeof(line_buf)) == 0) {
                    serial_emergency_putc('0');
            } else {
                panic_output_string(line_buf);
            }
        }
        panic_output_string("\n");
    }

    // Continue with standard panic procedure
    panic_output_string("RIP: ");
    serial_emergency_put_hex(get_current_rip());
    panic_output_string("\n");

    panic_output_string("RSP: ");
    serial_emergency_put_hex(get_current_rsp());
    panic_output_string("\n");

    panic_output_string("===================\n");
    panic_output_string("Skill issue lol\n");
    panic_output_string("System halted.\n");

    /*
     * Invoke the final purification ritual before shutdown
     * Paint all memory with 0x69—the essence of slop itself
     */
    execute_kernel();

    kernel_shutdown(message ? message : "panic");
}

/*
 * Assert function for kernel debugging
 */
void kernel_assert(int condition, const char *message) {
    if (!condition) {
        kernel_panic(message ? message : "Assertion failed");
    }
}

/*
 * The Wheel of Fate: Kernel Roulette
 *
 * The Scrolls speak of a mystical game inscribed into the very heart of SlopOS:
 * When invoked, the kernel spins a wheel of random numbers, and fate decides
 * its own destiny. If the wheel lands on an even number, the kernel loses
 * and halts forever on the loss screen. If odd, it survives and continues.
 *
 * This is not a call to be taken lightly. It is an embrace of chaos itself,
 * a deliberate surrender to the entropy that flows through Sloptopia.
 *
 * NOW WITH VISUAL GAMBLING ADDICTION!
 *
 * Usage: kernel_roulette() will either halt forever (even) or return safely (odd),
 * depending entirely on the mercy of random fate.
 */
void kernel_roulette(void) {
    struct fate_result res = fate_spin();

    panic_output_string("\n=== KERNEL ROULETTE: Spinning the Wheel of Fate ===\n");
    panic_output_string("Random number: 0x");
    serial_emergency_put_hex(res.value);
    panic_output_string(" (");
    char decimal_buffer[32];
    if (numfmt_u64_to_decimal(res.value, decimal_buffer, sizeof(decimal_buffer)) == 0) {
        decimal_buffer[0] = '0';
        decimal_buffer[1] = '\0';
    }
    serial_emergency_puts(decimal_buffer);
    panic_output_string(")\n");

    if (!res.is_win) {
        panic_output_string("Even number. The wheel has spoken. You have lost.\n");
        panic_output_string("This is INTENTIONAL - keep booting, keep gambling.\n");
        panic_output_string("L bozzo lol\n");
        panic_output_string("=== ROULETTE LOSS: AUTO-REBOOTING TO TRY AGAIN ===\n");
        panic_output_string("The gambling never stops...\n");
    } else {
        panic_output_string("Odd number. Fortune smiles upon the slop. Kernel survives.\n");
        panic_output_string("=== ROULETTE WIN: CONTINUING TO OS ===\n");
    }

    fate_apply_outcome(&res, FATE_RESOLUTION_REBOOT_ON_LOSS);
}

File: boot/src/kernel_panic.rs
use core::ffi::c_char;

use slopos_drivers::keyboard::keyboard_poll_wait_enter;
use slopos_drivers::serial;
use slopos_lib::cpu;
use slopos_lib::klog_info;
use slopos_lib::string::cstr_to_str;
use slopos_video::panic_screen;

use crate::shutdown::kernel_shutdown;

use crate::shutdown::execute_kernel;
use slopos_mm::memory_init::is_memory_system_initialized;

fn panic_output_str(s: &str) {
    serial::write_line(s);
}

fn read_rsp() -> u64 {
    let rsp: u64;
    unsafe {
        core::arch::asm!("mov {}, rsp", out(reg) rsp, options(nomem, nostack, preserves_flags));
    }
    rsp
}

#[derive(Clone, Copy)]
enum ControlRegister {
    Cr0,
    Cr3,
    Cr4,
}

fn read_cr(reg: ControlRegister) -> u64 {
    let value: u64;
    unsafe {
        match reg {
            ControlRegister::Cr0 => {
                core::arch::asm!("mov {}, cr0", out(reg) value, options(nomem, nostack, preserves_flags))
            }
            ControlRegister::Cr3 => {
                core::arch::asm!("mov {}, cr3", out(reg) value, options(nomem, nostack, preserves_flags))
            }
            ControlRegister::Cr4 => {
                core::arch::asm!("mov {}, cr4", out(reg) value, options(nomem, nostack, preserves_flags))
            }
        }
    }
    value
}

fn log_register_snapshot(rip: Option<u64>, rsp: Option<u64>) {
    panic_output_str("Register snapshot:");
    if let Some(rip) = rip {
        klog_info!("RIP: 0x{:x}", rip);
    }
    if let Some(rsp) = rsp {
        klog_info!("RSP: 0x{:x}", rsp);
    }
    klog_info!("CR0: 0x{:x}", read_cr(ControlRegister::Cr0));
    klog_info!("CR3: 0x{:x}", read_cr(ControlRegister::Cr3));
    klog_info!("CR4: 0x{:x}", read_cr(ControlRegister::Cr4));
}
pub fn kernel_panic(message: *const c_char) -> ! {
    cpu::disable_interrupts();

    panic_output_str("\n\n=== KERNEL PANIC ===");

    let msg_str = if !message.is_null() {
        let s = unsafe { cstr_to_str(message) };
        klog_info!("PANIC: {}", s);
        Some(s)
    } else {
        panic_output_str("PANIC: No message provided");
        None
    };

    let rsp = read_rsp();
    let cr0 = read_cr(ControlRegister::Cr0);
    let cr3 = read_cr(ControlRegister::Cr3);
    let cr4 = read_cr(ControlRegister::Cr4);
    log_register_snapshot(None, Some(rsp));

    panic_output_str("===================");
    panic_output_str("Kernel panic: unrecoverable error");

    // Display graphical panic screen if framebuffer is available
    if panic_screen::display_panic_screen(msg_str, None, Some(rsp), cr0, cr3, cr4) {
        panic_output_str("Press ENTER to shutdown...");
        keyboard_poll_wait_enter();
    } else {
        panic_output_str("System halted.");
    }

    if is_memory_system_initialized() != 0 {
        execute_kernel();
    } else {
        panic_output_str("Memory system unavailable; skipping paint ritual");
    }

    let reason = if message.is_null() {
        b"panic\0".as_ptr() as *const c_char
    } else {
        message
    };
    kernel_shutdown(reason);
}
pub fn kernel_panic_with_context(
    message: *const c_char,
    function: *const c_char,
    file: *const c_char,
    line: i32,
) {
    cpu::disable_interrupts();

    panic_output_str("\n\n=== KERNEL PANIC ===");

    let msg_str = if !message.is_null() {
        let s = unsafe { cstr_to_str(message) };
        klog_info!("PANIC: {}", s);
        Some(s)
    } else {
        None
    };

    unsafe {
        if !function.is_null() {
            klog_info!("Function: {}", cstr_to_str(function));
        }
        if !file.is_null() {
            klog_info!("File: {}:{}", cstr_to_str(file), line);
        }
    }

    let rsp = read_rsp();
    let cr0 = read_cr(ControlRegister::Cr0);
    let cr3 = read_cr(ControlRegister::Cr3);
    let cr4 = read_cr(ControlRegister::Cr4);
    log_register_snapshot(None, Some(rsp));

    panic_output_str("===================");
    panic_output_str("Kernel panic: unrecoverable error");

    // Display graphical panic screen if framebuffer is available
    if panic_screen::display_panic_screen(msg_str, None, Some(rsp), cr0, cr3, cr4) {
        panic_output_str("Press ENTER to shutdown...");
        keyboard_poll_wait_enter();
    } else {
        panic_output_str("System halted.");
    }

    if is_memory_system_initialized() != 0 {
        execute_kernel();
    } else {
        panic_output_str("Memory system unavailable; skipping paint ritual");
    }

    let reason = if message.is_null() {
        b"panic\0".as_ptr() as *const c_char
    } else {
        message
    };
    kernel_shutdown(reason);
}

pub fn kernel_panic_with_state(message: *const c_char, rip: u64, rsp: u64) {
    cpu::disable_interrupts();

    panic_output_str("\n\n=== KERNEL PANIC ===");

    let msg_str = if !message.is_null() {
        let s = unsafe { cstr_to_str(message) };
        klog_info!("PANIC: {}", s);
        Some(s)
    } else {
        panic_output_str("PANIC: No message provided");
        None
    };

    let cr0 = read_cr(ControlRegister::Cr0);
    let cr3 = read_cr(ControlRegister::Cr3);
    let cr4 = read_cr(ControlRegister::Cr4);
    log_register_snapshot(Some(rip), Some(rsp));

    panic_output_str("===================");
    panic_output_str("Kernel panic: unrecoverable error");

    // Display graphical panic screen if framebuffer is available
    if panic_screen::display_panic_screen(msg_str, Some(rip), Some(rsp), cr0, cr3, cr4) {
        panic_output_str("Press ENTER to shutdown...");
        keyboard_poll_wait_enter();
    } else {
        panic_output_str("System halted.");
    }

    if is_memory_system_initialized() != 0 {
        execute_kernel();
    } else {
        panic_output_str("Memory system unavailable; skipping paint ritual");
    }

    let reason = if message.is_null() {
        b"panic\0".as_ptr() as *const c_char
    } else {
        message
    };
    kernel_shutdown(reason);
}
pub fn kernel_assert(condition: i32, message: *const c_char) {
    if condition == 0 {
        let msg = if message.is_null() {
            b"Assertion failed\0".as_ptr() as *const c_char
        } else {
            message
        };
        kernel_panic(msg);
    }
}
